File: MobileAppReactNativeHooks/src/business/state/CashBalance.ts
import { print, round } from '../utils/utils';

class CashBalance {
  readonly amount: number;

  constructor(amount: number) {
    this.amount = isNaN(amount) ? 0 : round(amount);
  }

  public withAmount(amount: number): CashBalance {
    return new CashBalance(round(amount));
  }

  public add(howMuch: number): CashBalance {
    let newAmount = round(this.amount + howMuch);
    print(`Added ${howMuch}. Cash balance is now: ${newAmount}.`);
    return new CashBalance(newAmount);
  }

  public remove(howMuch: number): CashBalance {
    let newAmount = round(this.amount - howMuch);
    if (newAmount < 0) newAmount = 0;
    print(`Removed ${howMuch}. Cash balance is now: ${newAmount}.`);
    return new CashBalance(newAmount);
  }

  public toString(): string {
    return `US$ ${this.amount.toFixed(2)}`;
  }

  public equals(other: CashBalance): boolean {
    return (this === other) || (this.amount === other.amount);
  }
}

export default CashBalance;

File: MobileAppFlutterCelest/celest/lib/my_src/models/cash_balance.dart
import 'package:celest_backend/models.dart';
import 'package:celest_backend/my_src/models/utils/json.dart';
import 'package:meta/meta.dart';

@immutable
class CashBalance {
  static const CashBalance ZERO = CashBalance._(0);

  final double amount;

  CashBalance(double amount) : amount = amount.isNaN ? 0 : round(amount);

  const CashBalance._(this.amount);

  CashBalance withAmount(double amount) => CashBalance(round(amount));

  CashBalance add(double howMuch) {
    double newAmount = round(amount + howMuch);
    print('Added $howMuch. Cash balance is now: $newAmount.');
    return CashBalance(newAmount);
  }

  CashBalance remove(double howMuch) {
    double newAmount = round(amount - howMuch);
    if (newAmount < 0) newAmount = 0;
    print('Removed $howMuch. Cash balance is now: $newAmount.');
    return CashBalance(newAmount);
  }

  @override
  String toString() => 'US\$ ${amount.toStringAsFixed(2)}';

  Object? toJsonPersistor() => serialize<CashBalance>(this);

  factory CashBalance.fromJsonPersistor(Object? value) => deserialize<CashBalance>(value);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CashBalance && runtimeType == other.runtimeType && amount == other.amount;

  @override
  int get hashCode => amount.hashCode;
}
